internal/worker/client_test.go

```diff
@@ -101,39 +108,171 @@ func TestOAuth(t *testing.T) {
+		expectedCalls       []callDetails
+	}{
 
-	// initialize a test proxy server
-	proxy := &proxy{}
-	proxySrv := httptest.NewServer(proxy)
-	t.Cleanup(proxySrv.Close)
+		// we expect 5 or 7 calls to go through the proxy
+		// depending, if we "loose" authentication or not
 
-	client, err := worker.NewClient(worker.ClientConfig{
+		{"Test normal startup", false, []callDetails{
+			oauthCall,
+			registerCall,
+			requestJobCall,
+			uploadArtifactCall,
+			cancelBuildCall,
+		}},
+		{"Test loosing authentication", true, []callDetails{
+			oauthCall,
+			registerCall,
+			requestJobCall, // authentication lost - this call fails
+			oauthCall,
+			requestJobCall,
+			uploadArtifactCall,
+			cancelBuildCall,
+		}},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			muteLogrusDuringTest(t)
+
+			logrus.AddHook(&LogCheckerHook{
+				t:         t,
+				checkMsgs: []string{"Worker (x86_64) registered"},
+			})
+
+			workerURL, oauthURL, offlineToken := newTestWorkerServer(t)
+
+			// initialize a test proxy server
+			proxy := &proxy{registrationSuccessful: false}
+			proxySrv := httptest.NewServer(proxy)
+			t.Cleanup(proxySrv.Close)
+
+			client, err := worker.NewClient(worker.ClientConfig{
+				BaseURL:      workerURL,
+				TlsConfig:    nil,
+				ClientId:     "rhsm-api",
+				OfflineToken: offlineToken,
+				OAuthURL:     oauthURL,
+				BasePath:     "/api/image-builder-worker/v1",
+				ProxyURL:     proxySrv.URL,
+			})
+
+			if tc.resetAuthentication {
+				client.InvalidateAccessToken()
+			}
+
+			require.NoError(t, err)
+			job, err := client.RequestJob([]string{worker.JobTypeOSBuild}, "arch")
+			require.NoError(t, err)
+			r := strings.NewReader("artifact contents")
+			require.NoError(t, job.UploadArtifact("some-artifact", r))
+			c, err := job.Canceled()
+			require.False(t, c)
+			require.NoError(t, err)
+
+			require.Equal(t, len(proxy.paths), len(proxy.calls), "Error in test execution all calls should get a response")
+
+			for i, call := range tc.expectedCalls {
+				require.True(t, call.Equals(t, proxy.calls[i]), "Call idx %d does not match the expected sequence (got %v)", i, proxy.calls[i])
+			}
+
+			require.Equal(t, len(tc.expectedCalls), len(proxy.calls), "There are more calls than expected:\n"+strings.Join(proxy.paths, "\n"))
+
+			require.True(t, proxy.registrationSuccessful)
+		})
+	}
+}
+
+type LogCheckerHook struct {
+	checkMsgs []string
+	t         *testing.T
+	counter   int
+}
+
+func (h *LogCheckerHook) Levels() []logrus.Level {
+	return []logrus.Level{
+		logrus.DebugLevel,
+		logrus.InfoLevel,
+		logrus.WarnLevel,
+		logrus.ErrorLevel,
+		logrus.FatalLevel,
+		logrus.PanicLevel,
+	}
+}
+
+func (h *LogCheckerHook) Fire(e *logrus.Entry) error {
+	require.Less(h.t, h.counter, len(h.checkMsgs), "Got %s, probably unexpected", e.Message)
+	require.Contains(h.t, e.Message, h.checkMsgs[h.counter])
+	h.counter += 1
+	return nil
+}
+
+func muteLogrusDuringTest(t *testing.T) {
```

Comment:
Is this actually needed? Without `-v` go test eats all the stdout and stderr I think.